cpp/distributed/dist-mnist.cpp

```diff
@@ -0,0 +1,201 @@
+#include <torch/torch.h>
+#include <iostream>
+#include "mpi.h"
+
+std::map<at::ScalarType, MPI_Datatype> mpiDatatype = {
+        {at::kByte, MPI_UNSIGNED_CHAR},
+        {at::kChar, MPI_CHAR},
+        {at::kDouble, MPI_DOUBLE},
+        {at::kFloat, MPI_FLOAT},
+        {at::kInt, MPI_INT},
+        {at::kLong, MPI_LONG},
+        {at::kShort, MPI_SHORT},
+};
+
+//Define a Convolutional Module
+struct Model : torch::nn::Module {
+    Model()
+          : conv1(torch::nn::Conv2dOptions(1, 10, 5)),
+            conv2(torch::nn::Conv2dOptions(10, 20, 5)),
+            fc1(320, 50),
+            fc2(50, 10) {
+        register_module("conv1", conv1);
+        register_module("conv2", conv2);
+        register_module("conv2_drop", conv2_drop);
+        register_module("fc1", fc1);
+        register_module("fc2", fc2);
+    }
+
+    torch::Tensor forward(torch::Tensor x) {
+        x = torch::relu(torch::max_pool2d(conv1->forward(x), 2));
+        x = torch::relu(torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));
+        x = x.view({-1, 320});
+        x = torch::relu(fc1->forward(x));
+        x = torch::dropout(x, 0.5, is_training());
+        x = fc2->forward(x);
+        return torch::log_softmax(x, 1);
+    }
+
+    torch::nn::Conv2d conv1;
+    torch::nn::Conv2d conv2;
+    torch::nn::Dropout2d conv2_drop;
+    torch::nn::Linear fc1;
+    torch::nn::Linear fc2;
+};
+
+
+int main(int argc, char *argv[])
+{
+
+  //MPI variables
+  int rank, numranks;
+  MPI_Init(&argc, &argv);
+  MPI_Comm_size(MPI_COMM_WORLD, &numranks);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  //end MPI variables
+
+  // Timer variables
+  auto tstart = 0.0;
+  auto tend = 0.0;
+
+  //TRAINING
+  //Read train dataset
+  std::string filename = "/afs/crc.nd.edu/user/s/sghosh2/Public/ML/mnist/data";
```

Comment:
Should we make this an input arg instead of hard-coding a path?

Reply:
Forgot to remove that! Should I put the path as a user input or search for it in a subfolder as shown in this [example](https://github.com/pytorch/examples/blob/master/cpp/mnist/mnist.cpp) ?

Reply:
changed to "../data"